Update README.md (#26)
cozy-hosting/snug: README.md

# Snug

Snug ist ein Commandline Tool mit integrierter Skriptsprache für die **cozy** Cloud Plattform. Die Skriptsprache ist auf
Basis einer internen Kotlin DSL errichtet und bietet die Möglichkeit, komplexe Deployments in Code umzusetzen 
(Infrastructure as Code).

Ein Beispiel für den Einsatz der Skripsprache stellen Szenarien in einer CI/CD Pipeline dar, bei denen ein Skript dazu
dienen kann die Infrastruktur ensprechend des veränderten Codes anzupassen, oder neu bereitzustellen. 
Das Skript lässt sich hierbei aufgrund seiner Basis in Code leicht versionieren.

Darüber hinaus können mit den Skripten auch im Zuge der Plattformentwicklung einfach Templates für *one-click*
Deployments erzeugt sowie ebenfalls von den Betreibern beziehungsweise auch den Kunden selbst herausgegeben und
versioniert werden.

> Das initiale Release konzentriert sich derzeit in größerem Umfang auf die Skriptsprache, als das Commandline Tool.
> Künftig soll aber auch eine Steuerung der Plattform Teil von Snug werden und die Sprache nur ein Modul
> des gesammten Tools sein.

> > Darüberhinaus ist die Skripsrache derzeit ein *Proof of Concept*, der Produktiveinsatz ist noch nicht möglich.

Author: ```suggestion
> Darüber hinaus ist die Skripsrache derzeit ein *Proof of Concept*, der Produktiveinsatz ist noch nicht möglich.
```

## Skriptbeispiel

Das gezeigtes Beispiel stellt einen NGINX Webserver der neusten verfügbaren Version bereit. 

> Dieser ist auf Port `80`innerhalb des privaten Netzwerkes der Plattform erreichbar, noch nicht aber im Internet.

Author: ```suggestion
Dieser ist auf Port `80` innerhalb des privaten Netzwerkes der Plattform erreichbar, aber noch nicht im Internet.
```

```kotlin
deployment("nginx") {
    tags("personal", "website")
    replicas = 3

    container("webserver") {
        image("nginx", "latest")

        ports {
            "http" from 80 to 8080
        }
    }
}
```

## How to build?

### Docker

Möglichst einfach lässt sich Snug erzeugen, indem [Docker](https://www.docker.com/) verwendet wird. 

> Hierzu lediglich das Docker Image erstellt werden.

Author: ```suggestion
Hierzu muss lediglich das Docker Image erstellt werden.
```

| Publish                | Aggregat von PublishedPorts und PublishedDomains für ein Deployment         |
| PublishedPort          | Zuweisung eines Container Ports auf einen zufälligen öffentlichen Port      |
| PublishedDomains       | Zuweisung eines Container Ports auf eine bestimmte Domain                   |
| LoadBalancer           | Lastverteiler auf der Plattform                                             |
| LoadBalancedDeployment | Ein Deployment, das an einen LoadBalancer gekoppelt ist                     |
| Storage                | Abbildung eines Speichermediums mit StorageClass und Size auf der Plattform |
| StorageClass           | Die spezifische Art eines Speichermediums                                   |
| Size                   | Abbildung einer beliebigen Speichergröße normiert auf Gigabyte              |

###  Analyse und Begründung der verwendeten Muster
Snug verwendet *Aggregates*, *Entities*, *Value-Objects*, *Repositories* und *Services*.
Einige dieser gehen auch bereits aus dem vorangehenden Abschnitt [Ubiquitous Language](#ubiquitous-language) hervor.

> So gibt es beispielsweise *Aggregates*, wie etwa `Ressources`, das Skripte gebündelt bereitstellt, 

Author: ```suggestion
So gibt es beispielsweise *Aggregates*, wie etwa `Ressources`, welche Skripte gebündelt bereitstellt, 
```

um batch-prossesing zu erlauben.
*Entities* werden in Snug formuliert und sind Datenobjekte, sowie Bausteine der internen DSL.
Alle Datenobjekte in Snug sind zusätzlich *Value-Objects*.
Diese Basis ergibt sich aus den in Kotlin spezifischen `data class` Objekten.
*Repositories* kommen in Snug für klassische Persitenzaufgaben zum Einsatz.
So gibt es etwa das `ConfigRepository`, welches Abbilder der `Config` auf dem Dateisystem verwaltet.
Neben ihnen werden auch *Services* eingetzt.
Bei ihnen geht es um die allgemeine Bündelung von Funktionalitäten für einen bestimmten Bereich der Domäne.
Ein solch sezifischer Problembereich ist in Snug das Laden und Ausführung von Skripten.

> Hierzu existiert dann ein `ScriptService`, der Funktionalität zur bewältigung eben dieses Problem bereitstellt.

Author: ```suggestion
Hierzu existiert dann ein `ScriptService`, der Funktionalität zur Bewältigung eben dieses Problems bereitstellt.
```

# Snug

[![Build Pipeline](https://github.com/cozy-hosting/snug/actions/workflows/default.yml/badge.svg)](https://github.com/cozy-hosting/snug/actions/workflows/default.yml)
![Lines of code](https://img.shields.io/tokei/lines/github/cozy-hosting/snug)

> Snug ist ein Commandline Tool mit integrierter Skriptsprache für die **cozy** Cloud Plattform. Die Skriptsprache ist auf
> Basis einer internen Kotlin DSL errichtet und bietet die Möglichkeit, komplexe Deployments in Code umzusetzen 
> (Infrastructure as Code).

Author: ```suggestion
Snug ist ein Commandline Tool mit integrierter Skriptsprache für die **cozy** Cloud Plattform. Die Skriptsprache ist auf
Basis einer internerrichtet und bietet die Möglichkeit, komplexe Deployments in Code umzusetzen 
(Infrastructure as Code).
```

> des gesammten Tools sein.
> Darüberhinaus ist die Skripsrache derzeit ein *Proof of Concept*, der Produktiveinsatz ist noch nicht möglich.

## Skriptbeispiel

Das gezeigtes Beispiel stellt einen NGINX Webserver der neusten verfügbaren Version bereit. 
Dieser ist auf Port `80`innerhalb des privaten Netzwerkes der Plattform erreichbar, noch nicht aber im Internet.

```kotlin
deployment("nginx") {
    tags("personal", "website")
    replicas = 3

    container("webserver") {
        image("nginx", "latest")

        ports {
            "http" from 80 to 8080
        }
    }
}
```

## How to build?

### Docker

Möglichst einfach lässt sich Snug erzeugen, indem [Docker](https://www.docker.com/) verwendet wird. 
Hierzu lediglich das Docker Image erstellt werden.
Dies geschieht mit dem unten aufgeführten Befehl.

```shell
docker build -t snug:local <PROJECT ROOT>
```

Anschließend lässt sich die Anwendung auch per Docker Befehl starten.
Snug wird hierbei stateless gestartet.
Es werden also keine Zustände für den nächsten Aufruf gesichert.

```shell
docker run --rm snug:local [ARGS]
```

> 

Reviewer: Maybe provide a example apply of the example script here

| StorageClass           | Die spezifische Art eines Speichermediums                                   |
| Size                   | Abbildung einer beliebigen Speichergröße normiert auf Gigabyte              |

###  Analyse und Begründung der verwendeten Muster
Snug verwendet *Aggregates*, *Entities*, *Value-Objects*, *Repositories* und *Services*.
Einige dieser gehen auch bereits aus dem vorangehenden Abschnitt [Ubiquitous Language](#ubiquitous-language) hervor.

So gibt es beispielsweise *Aggregates*, wie etwa `Ressources`, das Skripte gebündelt bereitstellt, 
um batch-prossesing zu erlauben.
*Entities* werden in Snug formuliert und sind Datenobjekte, sowie Bausteine der internen DSL.
Alle Datenobjekte in Snug sind zusätzlich *Value-Objects*.
Diese Basis ergibt sich aus den in Kotlin spezifischen `data class` Objekten.
*Repositories* kommen in Snug für klassische Persitenzaufgaben zum Einsatz.
So gibt es etwa das `ConfigRepository`, welches Abbilder der `Config` auf dem Dateisystem verwaltet.
Neben ihnen werden auch *Services* eingetzt.
Bei ihnen geht es um die allgemeine Bündelung von Funktionalitäten für einen bestimmten Bereich der Domäne.

> Ein solch sezifischer Problembereich ist in Snug das Laden und Ausführung von Skripten.

Reviewer: Ein solch *spezifischer*

> Ein solch sezifischer Problembereich ist in Snug das Laden und Ausführung von Skripten.

Reviewer: ```suggestion
Ein solch spezifischer Problembereich ist in Snug das Laden und Ausführung von Skripten.
```
